worker/transcribee_worker/diarize.py

```diff
@@ -1,78 +1,42 @@
-
+def diarize(audio: npt.NDArray, doc: Document):
+    logging.info("Running diarization")
 
-def diarize(audio, progress_callback) -> List[Segment]:
-    pipeline = Pipeline.from_pretrained(
-        "pyannote/speaker-diarization@2.1",
-        use_auth_token=settings.HUGGINGFACE_TOKEN,
-        cache_dir=settings.MODELS_DIR,
-    )
-    device_name = "cpu"
-    if torch.backends.mps.is_available():
-        device_name = "mps"
-        pipeline.to("mps")
+    def time_to_sample(time: float | None):
+        if time is None:
+            raise ValueError("time may not be None")
+        return max(min(int(time / 1000 * settings.SAMPLE_RATE), len(audio)), 0)
 
-    def _hook(phase, *args, **kwargs):
-        total = kwargs.pop("total", None)
-        completed = kwargs.pop("completed", None)
-        phases = {
-            "segmentation": {"scale": 0.2, "offset": 0},
-            "embeddings": {"scale": 0.8, "offset": 0.2},
-        }
-        progress = None
-        if total is not None and completed is not None and phase in phases:
-            progress = (
-                phases[phase]["offset"] + (completed / total) * phases[phase]["scale"]
-            )
-        progress_callback(
-            progress=ensure_float(
-                progress
-            ),  # Sometimes a numpy value -> convert to native python type
-            extra_data={
-                "phase": phase,
-                "total": ensure_int(
-                    total
-                ),  # Sometimes a numpy value -> convert to native python type
-                "completed": ensure_int(
-                    completed
-                ),  # Sometimes a numpy value -> convert to native python type
-                "device": device_name,
-            },
+    segments = [
+        (
+            time_to_sample(child.children[0].start),
+            time_to_sample(child.children[-1].end),
         )
+        for child in doc.children
+    ]
 
-    logging.info("Running diarization")
-    diarization = pipeline(audio, hook=_hook)
-    logging.info(f"Diarization Result {diarization}")
+    model = SpeakerEmbeddingModel(
+        str(Path(__file__).parent / "pyannote-embedding.onnx")
+    )
+    embeddings = model.generate_embeddings(audio, segments)
 
-    segments = []
-    for turn, _, speaker in diarization.itertracks(yield_label=True):
-        segment = Segment(
-            start=turn.start * 1000,
-            end=turn.end * 1000,
-            speakers=[int(speaker.split("_")[1])],
-        )
-        segments.append(segment)
-    segments.sort(key=lambda x: x.start)
+    # the threshold is carefully hand-engineered through trail & error
+    # we might want a better protocol to define this by optimizing some metric :D
+    clustering = AgglomerativeClustering(
+        compute_full_tree=True, distance_threshold=1200, n_clusters=None
+    )
+    clustering.fit(np.array(embeddings))
 
-    return segments
+    for para, label in zip(doc.children, clustering.labels_):
```

Comment:
I think it would be nice to rename the speakers here so that `1` is the speaker that appears first in the document, `2` the one that appears second, ...